### PR TITLE
ARROW-5775: [C++] Fix thread-unsafe cached data

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
Hmm... so it looks like gcc 4.8.2 in manylinux1 may not have these APis:
https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

**Comment by pitrou** [comment]
Perhaps we can workaround that by emulating `atomic_load` and `atomic_store` on that old gcc using a global lock.

**Comment by pitrou** [comment]
Note that manylinux2010 works fine... one more reason to switch.

**Comment by pitrou** [comment]
Travis-CI: https://travis-ci.org/pitrou/arrow/builds/553762594

**Comment by pitrou** [comment]
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/25722036

File: cpp/src/arrow/util/atomic_shared_ptr.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+template <class T>
+void atomic_store(std::shared_ptr<T>* p, std::shared_ptr<T> r) {
+  *p = r;
+}
+
+#else
+
+template <class T>
+std::shared_ptr<T> atomic_load(const std::shared_ptr<T>* p) {
```
Comment (kszucs) [inline cpp/src/arrow/util/atomic_shared_ptr.h]: Inline?

Comment (pitrou) [reply cpp/src/arrow/util/atomic_shared_ptr.h]: Probably indeed. Will do.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -44,12 +44,16 @@ if(arrow_VERSION_MAJOR STREQUAL "0")
   set(ARROW_SO_VERSION "${ARROW_VERSION_MINOR}")
   set(ARROW_FULL_SO_VERSION "${ARROW_SO_VERSION}.${ARROW_VERSION_PATCH}.0")
 else()
-  message(FATAL_ERROR "Need to implement SO version generation for Arrow 1.0+")
+  message(WARNING "Need to implement SO version generation for Arrow 1.0+, "
+                  "using temporary pre-1.0 numbers")
+  set(ARROW_SO_VERSION "99")
+  set(ARROW_FULL_SO_VERSION "99.0.0")
```
Comment (pitrou) [inline cpp/CMakeLists.txt]: @kou Is that ok for you?

Comment (xhochy) [reply cpp/CMakeLists.txt]: I would prefer to simply go with https://github.com/apache/arrow/pull/4801

**Review by kszucs — APPROVED** [review]
@pitrou needs a rebase

**Comment by pitrou** [comment]
Done.

**Comment by kszucs** [comment]
Appveyor R failure is unrelated.

**Pull request closed without merging**